### PR TITLE
[ADAM-1695] Check for illegal genotype index after splitting multi-allelic variants.

### LINKED ISSUE #1695: Exception thrown in VariantContextConverter.formatAllelicDepth despite SILENT validation stringency
```
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/NA12878_HG001/NISTv3.3.2/GRCh38/supplementaryFiles/HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_all.vcf.gz

$ dsh-bio remap-phase-set \
  -i HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_all.vcf.gz \
  -o HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_all.fixed-phase-set.vcf.gz

$ adam-shell

import org.bdgenomics.adam.rdd.ADAMContext._

val variants = sc.loadVariants("HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_all.fixed-phase-set.vcf.gz")

val filtered = variants.transform(rdd => { rdd.filter(_.getFiltersPassed()) })

filtered.toVariantContextRDD().saveAsVcf("HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_all.fixed-phase.set.adam-filtered.vcf", asSingleFile = true, deferMerging = false, disableFastConcat = false, stringency = htsjdk.samtools.ValidationStringency.SILENT)

...
java.lang.ArrayIndexOutOfBoundsException: 3
	at org.bdgenomics.adam.converters.VariantContextConverter.formatAllelicDepth(VariantContextConverter.scala:779)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$40.apply(VariantContextConverter.scala:926)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$40.apply(VariantContextConverter.scala:926)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$77$$anonfun$apply$79.apply(VariantContextConverter.scala:1666)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$77$$anonfun$apply$79.apply(VariantContextConverter.scala:1666)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$78.apply(VariantContextConverter.scala:1668)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$78.apply(VariantContextConverter.scala:1668)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at org.bdgenomics.adam.converters.VariantContextConverter.org$bdgenomics$adam$converters$VariantContextConverter$$convert$2(VariantContextConverter.scala:1668)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$makeGenotypeFormatFn$1.apply(VariantContextConverter.scala:1708)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$makeGenotypeFormatFn$1.apply(VariantContextConverter.scala:1708)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$convert$3$$anonfun$6.apply(VariantContextConverter.scala:335)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$convert$3$$anonfun$6.apply(VariantContextConverter.scala:334)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$convert$3.apply(VariantContextConverter.scala:334)
	at org.bdgenomics.adam.converters.VariantContextConverter$$anonfun$convert$3.apply(VariantContextConverter.scala:324)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.bdgenomics.adam.converters.VariantContextConverter.convert(VariantContextConverter.scala:324)
	at org.bdgenomics.adam.rdd.ADAMContext$$anonfun$loadVcf$1$$anonfun$apply$14.apply(ADAMContext.scala:2017)
	at org.bdgenomics.adam.rdd.ADAMContext$$anonfun$loadVcf$1$$anonfun$apply$14.apply(ADAMContext.scala:2017)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:461)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439)
	at org.apache.spark.internal.io.SparkHadoopMapReduceWriter$$anonfun$4.apply(SparkHadoopMapReduceWriter.scala:146)
	at org.apache.spark.internal.io.SparkHadoopMapReduceWriter$$anonfun$4.apply(SparkHadoopMapReduceWriter.scala:144)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1375)
	at org.apache.spark.internal.io.SparkHadoopMapReduceWriter$.org$apache$spark$internal$io$SparkHadoopMapReduceWriter$$executeTask(SparkHadoopMapReduceWriter.scala:159)
	at org.apache.spark.internal.io.SparkHadoopMapReduceWriter$$anonfun$3.apply(SparkHadoopMapReduceWriter.scala:89)
	at org.apache.spark.internal.io.SparkHadoopMapReduceWriter$$anonfun$3.apply(SparkHadoopMapReduceWriter.scala:88)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by heuermh**
Fixes #1695

### CODE REVIEW

**Comment by fnothaft** [comment]
OOC, do you know what the license on the GIAB data is? We should ensure that it is OK to check in an excerpt of the GIAB truth file.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantContextConverter.scala [hunk 1]
```diff
@@ -787,7 +787,7 @@ class VariantContextConverter(
                                              gIndices: Array[Int]): Genotype.Builder = {
 
     // AD is an array type field
-    if (g.hasAD) {
+    if (g.hasAD && gIdx < g.getAD.size) {
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantContextConverter.scala]: Does this only affect the AD field? I would've expected it to broadly impact fields that have the same array count as `AD`? Do we already check those?

**Comment by heuermh** [comment]
> OOC, do you know what the license on the GIAB data is? We should ensure that it is OK to check in an excerpt of the GIAB truth file.

"License: There are no restrictions on the use of this data."

https://aws.amazon.com/public-datasets/giab

"[P]lease cite http://www.nature.com/nbt/journal/v32/n3/full/nbt.2835.html (doi:10.1038/nbt.2835) and http://www.nature.com/articles/sdata201625 (doi:10.1038/sdata.2016.25) when using these calls."

ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/NA12878_HG001/latest/README_NISTv3.3.2.txt

I can add an accompanying HG...vcf.README to the `adam-core/src/test/resources` directory.

**Comment by fnothaft** [comment]
> I can add an accompanying HG...vcf.README to the adam-core/src/test/resources directory.

+1

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantContextConverter.scala (hunk 1 above)
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantContextConverter.scala]: Yes, that file also has an `ADALL` VCF FORMAT Number=R Type=Integer genotype field that causes ArrayIndexOutOfBoundsExceptions, thus the other change at line 1289.

I'll add additional assertions to the test case to demonstrate this fix.

Comment (fnothaft) [reply adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantContextConverter.scala]: Oh, hah! I'd missed that change earlier. Thanks for clarifying!

**Comment by heuermh** [comment]
Rebased, ready for review @devin-petersohn

**Comment by devin-petersohn** [comment]
Merged, thanks @heuermh!

**Pull request merged**